### ISSUE TITLE
snapcraft.io pods crash if API calls to the snap store fail

### ISSUE THREAD

**Issue by axinojolais** [canonical-websites/snapcraft.io#329]
Hi,

We had a situation today where the snap store was unavailable. This made snapcraft.io display a 503 because the pods kept crashing. See https://pastebin.canonical.com/p/ff38P7VFFK/ and https://pastebin.canonical.com/p/Mcd79RdRwc/

Please make snapcraft.io handle failed API calls better please, thanks

**Comment by noise** [comment]
If possibly it would be good to hold a copy of last-good response for at least the needs of the homepage, and fallback to that when APIs are unresponsive or erroring.

**Comment by nottrobin** [comment]
@noise this is actually what we already do, but it failed because of the way the kubernetes deployment is configured. What happens if the API is down is that the API request will time out, at which point the last good response is returned from the cache, but this takes a number of seconds (not exactly sure how many at the moment).

The reason this crashes the units is because they're configured to do liveness checks that timeout in 1 second. So when the API is slow to respond, Kubernetes thinks the units aren't working and so restarts them, continually.

The fix, which we're already working on, is to:

- Create a health-check endpoint for the liveness checks to check which doesn't rely on any API calls (done in https://github.com/canonical-websites/snapcraft.io/pull/327 but not live)
- Configure the liveness checks to check that endpoint instead
- Increase the timeouts on the liveness checks to 3 seconds
- Reduce the timeout on API calls, so they fail faster to return the last good response

**Comment by noise** [comment]
I can look at our search API timings and give you some suggestions for lower timeouts.

**Comment by noise** [comment]
I think you'd be safe setting a timeout of 2s, 99%iles for the endpoints you hit are well within that.

**Comment by nottrobin** [comment]
@noise coincidentally, that's what it already was :)

**Comment by nottrobin** [comment]
We've adjusted the timeouts on the readiness and liveness checks, but unfortunately the cache was independently broken. We disabled it currently, as explained in https://wiki.canonical.com/IncidentReports/2018-02-23-snapcraft.io-account-info-breach.

This will be fixed by updating the caching module with https://github.com/canonical-webteam/get-feeds/issues/2, https://github.com/canonical-webteam/get-feeds/issues/3 and https://github.com/canonical-webteam/get-feeds/issues/4. We will then re-enable caching in snapcraft.io to use this new caching module, which will fix this issue.

**Comment by Lukewh** [comment]
@nottrobin can you update/ close this issue when you get chance please?

**Comment by nottrobin** [comment]
Yes this has been somewhat mitigated through various means:

- ensuring that API calls time out effectively
- making the liveness checks check [the status page](https://snapcraft.io/status) instead of the homepage
- increasing the number of replicas running the site to 9
- `cache-control: max-age=61` on all public pages

We'll also soon have a squid frontend in front of the site, to help with load. And when the API does timeout, the user should now see a much friendlier looking error page, with the site header and footer.

Closing this now.

**Comment by 4a6f656c** [comment]
@nottrobin - could we please reopen this, as I believe the situation is far from resolved:

- Each container is running talisker/gunicorn/flask with a sync worker class and a single worker, which means that they can only serve a single request at any time.
- The API requests are taking 12 seconds to time out when the API service is unavailable.
- Since requests are being handled sequentially, a request to `/status` will block behind another request.

In short, any outage of the API service will result in an outage for snapcraft.io.

**Comment by nottrobin** [comment]
I believe these have been resolved, by #874, #829 and #811, so I'm going to close this again.

@4a6f656c please let me know if you disagree

 If anyone encounters for 503s in general or other errors I suggest they open a new issue at this point.
